File: src/lib/payload.ts
import { z } from "zod";

const MAX_PAYLOAD_BYTES = 5 * 1024 * 1024;

export const printJobPayloadSchema = z.object({
  type: z.enum(["raw", "escpos", "pdf", "image"]),
  encoding: z.literal("base64"),
  protocol: z.enum(["escpos", "zpl", "tspl", "raw"]).optional(),
  peripherals: z.object({
    drawer: z.enum(["pin2", "pin5", "none"]).optional(),
    cutter: z.enum(["partial", "full", "none"]).optional(),
    buzzer: z.enum(["epson_pulse", "star_bel", "none"]).optional(),
  }).optional(),
  data: z.string().min(1).refine((value) => {
    if (value.length > (MAX_PAYLOAD_BYTES / 3) * 4 + 8) return false;
    try {
      const decoded = Buffer.from(value, "base64");
      if (decoded.length === 0 || decoded.length > MAX_PAYLOAD_BYTES) return false;
      return decoded.toString("base64") === value;
    } catch {
      return false;
    }
  }, { message: `payload.data must be valid base64 and decode to 1..${MAX_PAYLOAD_BYTES} bytes` }),
}).superRefine((payload, ctx) => {
  const decoded = Buffer.from(payload.data, "base64");
  const pdfSignature = Buffer.from("%PDF-");
  const jpegSignature = decoded.length >= 3 && decoded[0] === 0xff && decoded[1] === 0xd8 && decoded[2] === 0xff;
  const looksLikePdf = decoded.length >= pdfSignature.length && decoded.subarray(0, pdfSignature.length).equals(pdfSignature);

  if (payload.type === "pdf" && !looksLikePdf) {
    ctx.addIssue({ code: z.ZodIssueCode.custom, path: ["data"], message: "PDF payload must start with the %PDF- signature" });
  }
  if (payload.type === "image" && !jpegSignature) {
    ctx.addIssue({ code: z.ZodIssueCode.custom, path: ["data"], message: "Image payload must be a JPEG" });
  }
  if ((payload.type === "raw" || payload.type === "escpos") && looksLikePdf) {
    ctx.addIssue({ code: z.ZodIssueCode.custom, path: ["data"], message: "PDF bytes cannot be labeled as raw/escpos; provide a real byte-stream payload or convert explicitly" });
  }

  if (payload.type === "raw") {
    if (!payload.protocol) {
      ctx.addIssue({ code: z.ZodIssueCode.custom, path: ["protocol"], message: "protocol is required for raw payloads" });
    }
  }
  if (payload.type === "escpos") {
    if (!payload.protocol || payload.protocol !== "escpos") {
      ctx.addIssue({ code: z.ZodIssueCode.custom, path: ["protocol"], message: "protocol 'escpos' is required for escpos payloads" });
    }
  }
  if ((payload.type === "pdf" || payload.type === "image") && payload.protocol) {
    ctx.addIssue({ code: z.ZodIssueCode.custom, path: ["protocol"], message: "protocol is not applicable for pdf/image payloads" });
  }
  if (payload.peripherals) {
    const hasPeripherals = [payload.peripherals.drawer, payload.peripherals.cutter, payload.peripherals.buzzer]
      .some((mode) => Boolean(mode) && mode !== "none");
    if (hasPeripherals && payload.protocol !== "escpos") {
      ctx.addIssue({ code: z.ZodIssueCode.custom, path: ["peripherals"], message: "peripherals are only supported for escpos protocol" });
    }
  }
});

export type PrintJobPayload = z.infer<typeof printJobPayloadSchema>;

export function validatePrintJobPayload(payload: unknown): PrintJobPayload {
  return printJobPayloadSchema.parse(payload);
}

export function buildTestPrintPayload(printerName: string, agentName: string): PrintJobPayload {
  // Names are user-controlled: strip C0 controls (incl. ESC) so a crafted
  // name cannot inject command bytes into the diagnostic ticket.
  const safeName = String(printerName ?? "").replace(/[^\x20-\x7e]/g, "").slice(0, 60);
  const safeAgent = String(agentName ?? "").replace(/[^\x20-\x7e]/g, "").slice(0, 60);
  const lines = [
    "\x1b\x40",
    "Odoo Print Agent\n",
    "Test Print\n",
    `Printer: ${safeName}\n`,
    `Agent: ${safeAgent}\n`,
    "------------------------\n",
    "Connection OK\n",
    "------------------------\n\n\n",
    "\x1d\x56\x01",
  ].join("");

  return { type: "escpos", protocol: "escpos", encoding: "base64", data: Buffer.from(lines, "binary").toString("base64") };
}

const TEST_PAGE_BYTES = 4096;

function safeTestText(value: string): string {
  return value.replace(/[^\x20-\x7e]/g, "").slice(0, 60);
}

export function buildTestPdfPayload(printerName: string, agentName: string): string {
  const safeName = safeTestText(printerName);
  const safeAgent = safeTestText(agentName);
  const stamp = new Date().toISOString().replace("T", " ").slice(0, 19);

  const streamContent = [
    "BT",
    "/F1 18 Tf",
    "50 720 Td",
    "(ODOO PRINT GATEWAY TEST PAGE) Tj",
    "/F1 12 Tf",
    "0 -30 Td",
    `(Printer: ${safeName}) Tj`,
    "0 -20 Td",
    `(Agent: ${safeAgent}) Tj`,
    "0 -20 Td",
    `(Status: OK | ${stamp}) Tj`,
    "ET",
  ].join("\n");

  const streamLength = Buffer.byteLength(streamContent, "utf-8");

  const header = "%PDF-1.4\n";
  const obj1 = "1 0 obj\n<< /Type /Catalog /Pages 2 0 R >>\nendobj\n";
  const obj2 = "2 0 obj\n<< /Type /Pages /Kids [3 0 R] /Count 1 >>\nendobj\n";
  const obj3 = "3 0 obj\n<< /Type /Page /Parent 2 0 R /MediaBox [0 0 612 792] /Resources << /Font << /F1 4 0 R >> >> /Contents 5 0 R >>\nendobj\n";
  const obj4 = "4 0 obj\n<< /Type /Font /Subtype /Type1 /BaseFont /Helvetica >>\nendobj\n";
  const obj5 = `5 0 obj\n<< /Length ${streamLength} >>\nstream\n${streamContent}\nendstream\nendobj\n`;

  const off1 = Buffer.byteLength(header);
  const off2 = off1 + Buffer.byteLength(obj1);
  const off3 = off2 + Buffer.byteLength(obj2);
  const off4 = off3 + Buffer.byteLength(obj3);
  const off5 = off4 + Buffer.byteLength(obj4);
  const startxref = off5 + Buffer.byteLength(obj5);

  const pad = (n: number) => String(n).padStart(10, "0");

  const xref = [
    "xref",
    "0 6",
    "0000000000 65535 f ",
    `${pad(off1)} 00000 n `,
    `${pad(off2)} 00000 n `,
    `${pad(off3)} 00000 n `,
    `${pad(off4)} 00000 n `,
    `${pad(off5)} 00000 n `,
    "trailer",
    "<< /Size 6 /Root 1 0 R >>",
    "startxref",
    String(startxref),
    "%%EOF\n",
  ].join("\n");

  return header + obj1 + obj2 + obj3 + obj4 + obj5 + xref;
}

/**
 * Build a test ticket in the LANGUAGE THE PRINTER ACTUALLY SPEAKS. An
 * ESC/POS ticket sent to a ZPL printer is garbage, and the strict capability
 * model now rejects it — so the gateway (like the Odoo router's
 * route_test_page) selects the template from the printer's declared
 * protocol. Document transports (spooler/ipp) use a minimal printable PDF ticket.
 */
export function buildTestPrintPayloadForPrinter(
  printerName: string,
  agentName: string,
  printer: { protocol?: string | null; connectionType?: string | null; capabilities?: { supported_protocols?: string[] } | null },
): PrintJobPayload {
  const declared = (printer.protocol ?? "").toLowerCase();
  const conn = (printer.connectionType ?? "").toLowerCase();
  const supported = (printer.capabilities?.supported_protocols ?? []).map((p) => String(p).toLowerCase());
  const byteProto = ["escpos", "zpl", "tspl", "raw"].includes(declared)
    ? declared
    : (["escpos", "zpl", "tspl", "raw"] as const).find((p) => supported.includes(p)) ?? "";
  const name = safeTestText(printerName);
  const agent = safeTestText(agentName);
  const stamp = new Date().toISOString().replace("T", " ").slice(0, 19);

  if (byteProto === "escpos") {
    return buildTestPrintPayload(printerName, agentName);
  }
  if (byteProto === "zpl") {
    const zpl = [
      "^XA",
      "^FO50,50^A0N,36,36^FDODOO PRINT GATEWAY TEST PAGE^FS",
      "^FO50,100^GB700,2,2^FS",
      `^FO50,120^A0N,28,28^FDPrinter : ${name}^FS`,
      `^FO50,160^A0N,28,28^FDAgent   : ${agent}^FS`,
      `^FO50,200^A0N,28,28^FDStatus  : OK | ${stamp}^FS`,
      "^XZ",
    ].join("\n");
    if (Buffer.byteLength(zpl, "utf-8") > TEST_PAGE_BYTES) throw new Error("test page exceeds limit");
    return { type: "raw", protocol: "zpl", encoding: "base64", data: Buffer.from(zpl, "utf-8").toString("base64") };
  }
  if (byteProto === "tspl") {
    const tspl = [
      "SIZE 75 mm, 50 mm",
      "GAP 2 mm, 0 mm",
      "DIRECTION 1",
      "CLS",
      'TEXT 50,40,"3",0,1,1,"ODOO PRINT GATEWAY TEST PAGE"',
      `TEXT 50,80,"2",0,1,1,"Printer : ${name}"`,
      `TEXT 50,110,"2",0,1,1,"Agent   : ${agent}"`,
      `TEXT 50,140,"2",0,1,1,"Status  : OK | ${stamp}"`,
      "PRINT 1,1",
    ].join("\n");
    if (Buffer.byteLength(tspl, "utf-8") > TEST_PAGE_BYTES) throw new Error("test page exceeds limit");
    return { type: "raw", protocol: "tspl", encoding: "base64", data: Buffer.from(tspl, "utf-8").toString("base64") };
  }
  if (byteProto === "raw") {
    const raw = [
      "================================",
      "  ODOO PRINT GATEWAY TEST PAGE ",
      "================================",
      `Printer : ${name}`,
      `Agent   : ${agent}`,
      "Protocol: RAW",
      "--------------------------------",
      `Status  : OK | ${stamp}`,
      "================================\n\n\n",
    ].join("\n");
    return { type: "raw", protocol: "raw", encoding: "base64", data: Buffer.from(raw, "utf-8").toString("base64") };
  }

  const isDocumentTransport =
    ["spooler", "ipp", "ipps"].includes(conn) ||
    ["spooler", "ipp", "ipps"].includes(declared) ||
    supported.some((p) => ["pdf", "spooler", "ipp", "ipps"].includes(p));

  if (isDocumentTransport) {
    const pdf = buildTestPdfPayload(name, agent);
    return {
      type: "pdf",
      encoding: "base64",
      data: Buffer.from(pdf, "utf-8").toString("base64"),
    };
  }

  throw new Error("This printer transport has no supported test ticket format; print a real report to validate the path.");
}

File: agent/internal/payload/payload.go
// Package payload implements the strict print-job payload contract shared
// between the Cloud Print Gateway and the Windows agent.
package payload

import (
	"bytes"
	"encoding/base64"
	"fmt"
)

type Type string

const (
	TypeRaw    Type = "raw"
	TypeESCPOS Type = "escpos"
	TypePDF    Type = "pdf"
	TypeImage  Type = "image"
)

const EncodingBase64 = "base64"
const MaxPayloadBytes = 5 * 1024 * 1024

type Peripherals struct {
	Drawer string
	Cutter string
	Buzzer string
}

type Payload struct {
	Type        Type
	Protocol    string
	Data        []byte
	Peripherals Peripherals
}

func Parse(raw interface{}) (*Payload, error) {
	if raw == nil {
		return nil, fmt.Errorf("payload is missing")
	}
	m, ok := raw.(map[string]interface{})
	if !ok {
		return nil, fmt.Errorf("payload must be a JSON object")
	}

	typ, _ := m["type"].(string)
	switch Type(typ) {
	case TypeRaw, TypeESCPOS, TypePDF, TypeImage:
	default:
		if typ == "" {
			return nil, fmt.Errorf("payload.type is required")
		}
		return nil, fmt.Errorf("unsupported payload type %q (expected %q, %q, %q or %q)", typ, TypeRaw, TypeESCPOS, TypePDF, TypeImage)
	}

	protocol, _ := m["protocol"].(string)

	switch Type(typ) {
	case TypeRaw:
		if protocol == "" {
			return nil, fmt.Errorf("protocol is required for raw payloads")
		}
		switch protocol {
		case "raw", "escpos", "zpl", "tspl":
		default:
			return nil, fmt.Errorf("unsupported protocol %q for raw payload", protocol)
		}
	case TypeESCPOS:
		if protocol == "" {
			return nil, fmt.Errorf("protocol is required for escpos payloads")
		}
		if protocol != "escpos" {
			return nil, fmt.Errorf("protocol %q is incompatible with escpos payload", protocol)
		}
	case TypePDF, TypeImage:
		if protocol != "" {
			return nil, fmt.Errorf("protocol is not applicable for %s payloads", typ)
		}
	}

	encoding, _ := m["encoding"].(string)
	if encoding != EncodingBase64 {
		return nil, fmt.Errorf("unsupported payload encoding %q (only %q is supported)", encoding, EncodingBase64)
	}
	dataStr, _ := m["data"].(string)
	if dataStr == "" {
		return nil, fmt.Errorf("payload.data is required")
	}
	if len(dataStr) > (MaxPayloadBytes/3)*4+8 {
		return nil, fmt.Errorf("payload too large (exceeds %d byte limit)", MaxPayloadBytes)
	}
	decoded, err := base64.StdEncoding.DecodeString(dataStr)
	if err != nil {
		return nil, fmt.Errorf("payload.data is not valid base64: %w", err)
	}
	if len(decoded) == 0 || len(decoded) > MaxPayloadBytes {
		return nil, fmt.Errorf("payload exceeds %d byte limit", MaxPayloadBytes)
	}

	// Signature search window (64 bytes) mirrors pdf.go's
	// pdfHeaderSearchWindow: producers that emit a UTF-8 BOM or stray
	// leading whitespace must pass intake exactly when execution accepts
	// them. Anything beyond the window is irrelevant to format detection.
	head := decoded
	if len(head) > 64 {
		head = head[:64]
	}
	looksLikePDF := bytes.Contains(head, []byte("%PDF-"))
	looksLikeJPEG := len(decoded) >= 3 && decoded[0] == 0xff && decoded[1] == 0xd8 && decoded[2] == 0xff
	if Type(typ) == TypePDF && !looksLikePDF {
		return nil, fmt.Errorf("PDF payload must start with the %%PDF- signature")
	}
	if Type(typ) == TypeImage && !looksLikeJPEG {
		return nil, fmt.Errorf("image payload must be a JPEG")
	}
	if (Type(typ) == TypeRaw || Type(typ) == TypeESCPOS) && looksLikePDF {
		return nil, fmt.Errorf("PDF bytes cannot be labeled as raw/escpos")
	}

	var periph Peripherals
	if periphMap, ok := m["peripherals"].(map[string]interface{}); ok {
		if d, ok := periphMap["drawer"].(string); ok {
			switch d {
			case "pin2", "pin5", "none":
				periph.Drawer = d
			default:
				return nil, fmt.Errorf("invalid drawer mode %q", d)
			}
		}
		if c, ok := periphMap["cutter"].(string); ok {
			switch c {
			case "partial", "full", "none":
				periph.Cutter = c
			default:
				return nil, fmt.Errorf("invalid cutter mode %q", c)
			}
		}
		if b, ok := periphMap["buzzer"].(string); ok {
			switch b {
			case "epson_pulse", "star_bel", "none":
				periph.Buzzer = b
			default:
				return nil, fmt.Errorf("invalid buzzer mode %q", b)
			}
		}
	}

	// "none" is an INACTIVE setting, not an action: it must never count as
	// a configured peripheral (serializing inactive peripherals as active
	// semantic values would make e.g. {drawer:"none"} a zpl payload killer).
	active := func(v string) bool { return v != "" && v != "none" }
	hasPeripherals := active(periph.Drawer) || active(periph.Cutter) || active(periph.Buzzer)
	if hasPeripherals && protocol != "escpos" {
		return nil, fmt.Errorf("peripherals are only supported for escpos protocol")
	}
	if !hasPeripherals {
		periph = Peripherals{}
	}

	return &Payload{
		Type:        Type(typ),
		Protocol:    protocol,
		Data:        decoded,
		Peripherals: periph,
	}, nil
}
